### PR TITLE
Introduction text for the authorization framework

### PR DESCRIPTION
**Pull request by tardyp**
blockdiag looks like this:

image

### CODE REVIEW

File: master/docs/manual/cfg-www.rst [hunk 1]
```diff
@@ -476,17 +476,71 @@ Here is a configuration that is known to work (Apache 2.4.10 / Debian 8), direct
 Authorization rules
 ~~~~~~~~~~~~~~~~~~~
 
+The authorization framework in Buildbot is very generic and flexible.
+Drawback is that it is not very obvious for new comers.
```
Comment (Lekensteyn) [inline master/docs/manual/cfg-www.rst]: "newcomers"

File: master/docs/manual/cfg-www.rst [hunk 2]
```diff
@@ -476,17 +476,71 @@ Here is a configuration that is known to work (Apache 2.4.10 / Debian 8), direct
...
+
+      User  [shape = actor];
+      AuthenticatedUser  [shape = actor];
+      RoleMatcher [shape = diamond];
+      EndpointMatcher [shape = diamond];
+    }
+
+- Roles is a label that you give to a user.
+  It is similar to the notion of group, but with more flexibility.
+  For an example of how it is different to group, there is the ``owner`` role, which can be given to a user for a build that he is at the origin.
```
Comment (Lekensteyn) [inline master/docs/manual/cfg-www.rst]: I would expect a group to be a collection of users. Users can be a member of multiple groups.
I would expect a role to be similar to this notion of groups. What is the difference? Groups are fixed while roles can dynamically assigned?

File: master/docs/manual/cfg-www.rst [hunk 3]
```diff
@@ -476,17 +476,71 @@ Here is a configuration that is known to work (Apache 2.4.10 / Debian 8), direct
...
+
+- Endpoint matchers associates role requirements to REST API endpoints.
```
Comment (Lekensteyn) [inline master/docs/manual/cfg-www.rst]: "associate"

Maybe clarify the default policy when nothing matches? (reject by default, but there seems to be a special case when the list is empty (allow all). I see it is mentioned below, but for a global picture I think it is an important detail.

File: master/docs/manual/cfg-www.rst [hunk 4]
```diff
@@ -476,17 +476,71 @@ Here is a configuration that is known to work (Apache 2.4.10 / Debian 8), direct
...
+
+- Role matchers associates authenticated users to roles.
```
Comment (Lekensteyn) [inline master/docs/manual/cfg-www.rst]: "associate" (or "A role matcher associates ...")

File: master/docs/manual/cfg-www.rst [hunk 5]
```diff
@@ -476,17 +476,71 @@ Here is a configuration that is known to work (Apache 2.4.10 / Debian 8), direct
...
+
+.. py:class:: buildbot.www.authz.Authz(allowRules=[], roleMatcher=[], stringsMatcher=util.fnmatchStrMatcher)
+
+    :param allowRules: List of EndpointMatchers processed in order for each endpoint grant request.
+    :param roleMatcher: List of RoleMatchers
+    :param stringsMatcher: Selects algorithm used to make string comparaison.
```
Comment (Lekensteyn) [inline master/docs/manual/cfg-www.rst]: "to make [a] string comparison" -> "to match a role with a username"? (if that makes sense?)

File: master/docs/manual/cfg-www.rst [hunk 6]
```diff
@@ -476,17 +476,71 @@ Here is a configuration that is known to work (Apache 2.4.10 / Debian 8), direct
...
+       can be :py:class:`util.fnmatchStrMatcher` or :py:class:`util.reStrMatcher` from ``from buildbot.plugins import util``
+
+    :py:class:`Authz` needs to be configured in ``c['www']['authz']``
+
 Endpoint matchers
 +++++++++++++++++
 
 Endpoint matchers are responsible for creating rules to match REST endpoints, and requiring roles for them.
-The following sequence is implemented by each EndpointMatcher class
+Endpoint matchers are process in the order they are configured.
```
Comment (Lekensteyn) [inline master/docs/manual/cfg-www.rst]: "are processed in"

File: master/docs/manual/cfg-www.rst [hunk 7]
```diff
@@ -476,17 +476,71 @@ Here is a configuration that is known to work (Apache 2.4.10 / Debian 8), direct
...
+The first that matches stop the sequence unless ``defaultDeny=False`` is passed.
```
Comment (Lekensteyn) [inline master/docs/manual/cfg-www.rst]: What about:

    The first rule matching an endpoint will prevent further rules from being checked. To continue checking other rules when the result is `deny`, set `defaultDeny=False`.

File: master/docs/manual/cfg-www.rst (hunk 5 above)
Comment (sa2ajj) [inline master/docs/manual/cfg-www.rst]: and, as spellchecking suggested, "comparison"

File: master/docs/manual/cfg-www.rst (hunk 3 above)
Comment (tardyp) [inline master/docs/manual/cfg-www.rst]: there is no reject by default. It is allow by default.

File: master/docs/manual/cfg-www.rst [hunk 8]
```diff
@@ -476,17 +476,86 @@ Here is a configuration that is known to work (Apache 2.4.10 / Debian 8), direct
...
+
 Endpoint matchers
 +++++++++++++++++
 
 Endpoint matchers are responsible for creating rules to match REST endpoints, and requiring roles for them.
-The following sequence is implemented by each EndpointMatcher class
+Endpoint matchers are processed in the order they are configured.
+The first rule matching an endpoint will prevent further rules from being checked.
+To continue checking other rules when the result is `deny`, set `defaultDeny=False`.
+If not endpoint matcher matches, then the access is granted.
```
Comment (Lekensteyn) [inline master/docs/manual/cfg-www.rst]: "If no"

File: master/docs/manual/cfg-www.rst [hunk 9]
```diff
@@ -476,17 +476,86 @@ Here is a configuration that is known to work (Apache 2.4.10 / Debian 8), direct
...
+      AuthenticatedUser  [shape = actor];
+      RoleMatcher [shape = diamond];
+      EndpointMatcher [shape = diamond];
+    }
+
+- Roles is a label that you give to a user.
+
+  It is similar but different to the usual notion of group:
+
+  - a user can have several roles, and a role can be given to several users.
```
Comment (Lekensteyn) [inline master/docs/manual/cfg-www.rst]: Nit: "A user"

File: master/docs/manual/cfg-www.rst [hunk 10]
```diff
@@ -476,17 +476,86 @@ Here is a configuration that is known to work (Apache 2.4.10 / Debian 8), direct
...
+- Look if the users has the required role.
+
+Several endpoints matchers are currently implemented.
+If you need a very complex setup, you may need to implement your own endpoint matchers.
+In this case, you can look at the source code for detailed examples on how to write endpoint matchers.
+
+.. py:class:: buildbot.www.authz.endpointmatchers.EndpointMatcherBase(role, defaultDeny=True)
+
+    :param role: The role which grants access to this endpoint.
+        List of roles is not supported, but a ``fnmatch`` expression can be provided to match several roles.
```
Comment (Lekensteyn) [inline master/docs/manual/cfg-www.rst]: Isn't fnmatch only available when `stringsMatcher=fnmatchStrMatcher` (default)?

Comment (tardyp) [inline master/docs/manual/cfg-www.rst]: indeed. I did not want to make it more complicated than it needs

**Comment by sa2ajj** [comment]
After @Lekensteyn's issues are address, this is good to go.

**Comment by tardyp** [comment]
@Lekensteyn remarks should be addressed now

**Pull request merged**